crates/legacy/src/service.rs

```diff
@@ -1235,15 +1778,3188 @@ pub(crate) async fn handle_received_txns<Types: NodeType>(
+    #[async_std::test]
+    async fn test_get_available_blocks_error_get_channel_for_matching_builder() {
+        let (bootstrap_sender, _) = async_broadcast::broadcast(10);
+        let (tx_sender, _) = async_broadcast::broadcast(10);
+        let (builder_public_key, builder_private_key) =
+            <BLSPubKey as BuilderSignatureKey>::generated_from_seed_indexed([0; 32], 0);
+        let (leader_public_key, leader_private_key) =
+            <BLSPubKey as SignatureKey>::generated_from_seed_indexed([0; 32], 1);
+        let parent_commit = vid_commitment(&[], 8);
+
+        let state = Arc::new(ProxyGlobalState::<TestTypes>::new(
+            Arc::new(RwLock::new(GlobalState::<TestTypes>::new(
+                bootstrap_sender,
+                tx_sender,
+                parent_commit,
+                ViewNumber::new(4),
+                ViewNumber::new(4),
+                10,
+            ))),
+            (builder_public_key, builder_private_key.clone()),
+            Duration::from_secs(1),
+        ));
+
+        {
+            let mut write_locked_global_state = state.global_state.write_arc().await;
+            write_locked_global_state.highest_view_num_builder_id = BuilderStateId {
+                parent_commitment: parent_commit,
+                view: ViewNumber::new(5),
+            };
+        }
+
+        // As a result, we **should** be receiving a request for the available
+        // blocks with our expected state id on the receiver, along with a channel
+        // to send the response back to the caller.
+
+        let signature = BLSPubKey::sign(&leader_private_key, parent_commit.as_ref()).unwrap();
+        let result = state
+            .available_blocks_implementation(&parent_commit, 6, leader_public_key, &signature)
+            .await;
+        match result {
+            Err(AvailableBlocksError::GetChannelForMatchingBuilderError(_)) => {
+                // This is what we expect.
+                // This message *should* indicate that the response channel was closed.
+            }
+            Err(err) => {
+                panic!("Unexpected error: {:?}", err);
+            }
+            Ok(_) => {
+                panic!("Expected an error, but got a result");
+            }
+        }
+    }
+
+    // We have two error cases for `available_blocks_implementation` that we
+    // cannot seem trigger directly due to the nature of how the implementation
+    // performs.
+    //
+    // The first is ChannelUnexpectedlyClosed, which doesn't seem to be
+    // producible as the unbounded channel doesn't seem to be able to be
+    // closed.
+    //
+    // The second is SigningBlockFailed, which doesn't seem to be producible
+    // with a valid private key, and it's not clear how to create an invalid
+    // private key.
+
+    /// This test checks that call to `available_blocks_implementation` returns
+    /// a successful response when the function is called before blocks are
+    /// made available.
```

Comment:
I believe this may be true in general.  I wasn't certain at the time I wrote the test.  I do think if this assumption were relaxed and not codified with a test that we could potentially change this behavior so that it does affect this behavior.  However, I don't know how likely that would be.

Reply:
Actually, do we want `available_blocks_implementation` to return the blocks if it's called before blocks are made available? This test passes because we call `await` after the blocks are available, but if we change it to be really "before", won't we expect no available blocks to be returned? I feel it's a bit counter-intuitive to expect `available_blocks_implementation` to give the same result in both cases.

Reply:
Well, in this case we're dealing with non-synchronous processes.  The `available_blocks_implementation` does have a loop where it attempts to wait for a matching builder with a while loop using this variable: `time_to_wait_for_matching_builder`.  So it definitely seems like there's an expectation that we should be able to receive a request for the block, before producing the block and being able to fulfill this.

Due to the async nature of the API requests versus the receiving of the Proposal events I think this makes sense, as we cannot guarantee that by the time `HotShot` calls the `Builder`'s endpoint that we will have already receive a `Proposal` from the HotShot `Event Stream`.

Reply:
Makes sense! I missed the fact that `available_blocks_implementation` was wrapped by `async_spawn(async move {...})` so the task had been started before calling `await`. Agree that the order matters in this async case.